File: src/main/kotlin/unizar/labis/g03/backendapp/services/ReservarEspacioService.java
package unizar.labis.g03.backendapp.services;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import unizar.labis.g03.backendapp.model.DTO.ReservaDTO;
import unizar.labis.g03.backendapp.model.entities.Espacio;
import unizar.labis.g03.backendapp.model.entities.Persona;
import unizar.labis.g03.backendapp.model.entities.Reserva;
import unizar.labis.g03.backendapp.model.valueObjects.InfoReserva;
import unizar.labis.g03.backendapp.repositories.EspacioRepository;
import unizar.labis.g03.backendapp.repositories.PersonaRepository;
import unizar.labis.g03.backendapp.repositories.ReservaRepository;

import java.util.List;

@Service
public class ReservarEspacioService {
    private final ReservaRepository reservaRepository;
    private final EspacioRepository espacioRepository;
    private final PersonaRepository personaRepository;

    @Autowired
    public ReservarEspacioService(ReservaRepository reservaRepository, EspacioRepository espacioRepository, PersonaRepository personaRepository) {
        this.reservaRepository = reservaRepository;
        this.espacioRepository = espacioRepository;
        this.personaRepository = personaRepository;
    }

    //  Date fechaInicio, Date fechaFinal, int numAsistentesPrevistos, String descripcion
    public void reservarEspacios(ReservaDTO reservaDTO) {
        Reserva reserva = buildReserva(reservaDTO);
        InfoReserva infoReserva = reserva.getInfoReserva();
        for (Espacio espacio : reserva.getEspacios()) {
            if (!espacio.getReservable()) {
                //Espacio no reservable
            }
            if (espacio.getHorario().estaDentro(infoReserva.getHoraInicio(), infoReserva.getHoraFinal())) {
                // Horario no disponible
            }
            List<Reserva> reservasConflictivas = reservaRepository.encontrarReservasConflictivas
                    (espacio.getId(), infoReserva.getFechaInicio(), infoReserva.getFechaFinal());//Espacio no esta reservado
            if (!reservasConflictivas.isEmpty()) {
                //Si no es gerente
                if (reserva.getPersona().esGerente()) {
                    borrarReservasConflictivas(reservasConflictivas);
                    //save?
                }
            }
        }


        //Construimos el objeto reserva a partir del DTO recibido

        //Comprobamos que la reserva cumple con la politica de reservas

        //Guardamos la reserva en la base de datos
        reservaRepository.save(reserva);
    }

    private Reserva buildReserva(ReservaDTO reservaDTO) {
        List<Espacio> espacios = espacioRepository.findAllById(reservaDTO.getIdEspacios());
        Persona persona = personaRepository.findByEmail(reservaDTO.getEmailUsuario());
        InfoReserva infoReserva = new InfoReserva(reservaDTO.getNumAsistentesPrevistos(), reservaDTO.getFechaInicio(),
                reservaDTO.getFechaFinal(), reservaDTO.getDescripcion(), maximosOcupantesValido(espacios) );
        return new Reserva(persona, espacios, infoReserva);
    }

    private void borrarReservasConflictivas( List<Reserva> reservasConflictivas) {
        for (Reserva r : reservasConflictivas) {
            reservaRepository.anularReserva(r.getId());
        }
        //Espacio.opEspacio.maximosOcupantesValido()

    }

    private Integer  maximosOcupantesValido( List<Espacio> espacios){
        int suma = 0;
        for (Espacio espacio : espacios) {
            suma += espacio.getCapacidadMaxima();
        }
        return suma;
    }
}

File: src/main/kotlin/unizar/labis/g03/backendapp/application/services/ReservarEspacioService.kt
package unizar.labis.g03.backendapp.application.services

import jakarta.transaction.Transactional
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import unizar.labis.g03.backendapp.application.exceptions.EspacioYaReservadoException
import unizar.labis.g03.backendapp.application.exceptions.ReservaNoValidaException
import unizar.labis.g03.backendapp.application.exceptions.UsuarioNoEncontradoException
import unizar.labis.g03.backendapp.domain.model.DTO.ReservaDTO
import unizar.labis.g03.backendapp.domain.model.entities.Reserva
import unizar.labis.g03.backendapp.domain.model.valueObjects.InfoReserva
import unizar.labis.g03.backendapp.domain.repositories.EspacioRepository
import unizar.labis.g03.backendapp.domain.repositories.PersonaRepository
import unizar.labis.g03.backendapp.domain.repositories.ReservaRepository
import unizar.labis.g03.backendapp.domain.model.valueObjects.CalculoOcupantesMaximos
import unizar.labis.g03.backendapp.domain.services.ComprobarValidezReservas
import unizar.labis.g03.backendapp.domain.services.NotificarPersonas
import java.util.*

@Service
class ReservarEspacioService @Autowired constructor(
    private val reservaRepository: ReservaRepository,
    private val espacioRepository: EspacioRepository,
    private val personaRepository: PersonaRepository,
    private val comprobarValidezReservas: ComprobarValidezReservas,
    private val notificarPersonas: NotificarPersonas

) {
    companion object {
        private val ESPACIO_NO_DISPONIBLE = "El espacio ya esta reservado a la hora seleccionada"
    }
    @Transactional
    fun reservarEspacios(reservaDTO: ReservaDTO): Optional<Reserva> {
        val reserva = buildReserva(reservaDTO)
        val reservasConflictivas: MutableList<Reserva> = mutableListOf()
        val conflictos = comprobarValidezReservas.comprobar(reserva)
        if (conflictos != "") {
            throw ReservaNoValidaException(conflictos)
        }
        for (espacio in reserva.espacios) {
            reservasConflictivas.addAll(reservaRepository.encontrarReservasConflictivas(
                espacio.getId(), reserva.infoReserva.fechaInicio, reserva.infoReserva.fechaFinal)) //Espacio no esta reservado
            if (reservasConflictivas.isNotEmpty() && !reserva.persona.esGerente()) {
               throw EspacioYaReservadoException(ESPACIO_NO_DISPONIBLE)
            }
        }
        //Borramos las reservas conflictivas si existen.
        if(reservasConflictivas.isNotEmpty()){
            print("Reservas conflictivas: " + reservasConflictivas.get(0).toString())
            anularReservasConflictivas(reservasConflictivas)
        }
        reservaRepository.save(reserva)
        return Optional.of(reserva)
    }

    private fun buildReserva(reservaDTO: ReservaDTO): Reserva {
        val espacios = espacioRepository.findAllById(reservaDTO.getIdEspacios())
        val persona = personaRepository.findByEmail(reservaDTO.getEmailUsuario())
        if (persona.isEmpty) throw UsuarioNoEncontradoException("No existe usuario de la reserva")
        val infoReserva = InfoReserva(
            reservaDTO.getFechaInicio(), reservaDTO.getFechaFinal(), reservaDTO.getDescripcion(),
            CalculoOcupantesMaximos().calculo(espacios),
            reservaDTO.getNumAsistentesPrevistos(), reservaDTO.getTipoUso()
        )
        return Reserva(null, persona.get(), espacios, infoReserva)
    }

    private fun anularReservasConflictivas(reservasConflictivas: List<Reserva>) {
        val idsReservas: List<Int?> = reservasConflictivas.map { it.id }
        reservasConflictivas.forEach { notificarPersonas.notificaAnulacion(it) }
        reservaRepository.anularReservas(idsReservas)
    }

}
